src/SpectatorView.Unity/Assets/SpectatorView/Scripts/StateSynchronization/AssetCache.cs

```diff
@@ -178,83 +220,158 @@ public static IEnumerable<T> EnumerateAllComponentsInScenesAndPrefabs<T>()
+                }
+
+                if (lookupByAssetId.TryGetValue(assetId, out var entry))
                 {
-                    return assetEntry.Asset;
+                    return (TAsset)entry.Asset;
                 }
-                else
+
+                if (!TryLoadAssets(assetId.Name))
                 {
                     return default(TAsset);
                 }
+
+                if (!lookupByAssetId.TryGetValue(assetId, out entry))
+                {
+                    Debug.LogError($"Assets were loaded for {assetId.Name} but no associated asset cache entry was found for {assetId}");
+                    return default(TAsset);
+                }
+
+                return (TAsset)entry.Asset;
             }
         }
 
         public AssetId GetAssetId(TAsset asset)
         {
             using (StateSynchronizationPerformanceMonitor.Instance.IncrementEventDuration(this.GetType().Name, "GetAssetId"))
             {
-                TAssetEntry assetEntry;
-                if (asset != null && LookupByAsset.TryGetValue(asset, out assetEntry))
+                if (asset == null ||
+                    asset.name == null ||
+                    asset.name == string.Empty)
                 {
-                    return assetEntry.AssetId;
+                    return AssetId.Empty;
                 }
-                else
+
+                if (lookupByAsset.TryGetValue(asset, out var entry))
+                {
+                    return entry.AssetId;
+                }
+
+                if (!TryLoadAssets(asset.name))
                 {
                     return AssetId.Empty;
                 }
+
+                if (!lookupByAsset.TryGetValue(asset, out entry))
+                {
+                    Debug.LogError($"Assets were loaded for {asset.name} but not associated asset cache entry was found for {asset}");
+                    return AssetId.Empty;
+                }
+
+                return entry.AssetId;
+            }
+        }
+
+        public override void LoadAllAssets()
+        {
+            using (StateSynchronizationPerformanceMonitor.Instance.IncrementEventDuration(this.GetType().Name, "LoadAllAssets"))
+            {
+                bool successful = true;
+                foreach (var pair in LookupByName)
+                {
+                    if (!TryLoadAssets(pair.Key))
+                    {
+                        successful = false;
+                    }
+                }
+
+                if (!successful)
+                {
+                    Debug.LogWarning($"Not all assets were found for {this.GetType().Name} during forced asset loading");
+                }
+            }
+        }
+
+        private bool TryLoadAssets(string name)
+        {
+            using (StateSynchronizationPerformanceMonitor.Instance.MeasureEventMemoryUsage(this.GetType().Name, "LoadingAssets"))
+            {
+                if (!LookupByName.TryGetValue(name, out var assetsIds))

```

Comment:
>if (!LookupByName.TryGetValue(name, out var assetsIds)) [](start = 16, length = 55)

From a calling expectation perspective - if this has failed, is this actually an expected case, or does this reflect a failure on the caller (or something else earlier) (i.e. is it worth logging that an invalidly named asset/not found asset was attempted to be found?)

Reply:
This likely warrants an error.

---
In reply to: [355042828](https://github.com/microsoft/MixedReality-SpectatorView/pull/295#discussion_r355042828) [](ancestors = 355042828)